${config:} variable substitution in tasks.json

<!-- ⚠️⚠️ Do Not Delete This! bug_report_template ⚠️⚠️ -->
<!-- Please read our Rules of Conduct: https://opensource.microsoft.com/codeofconduct/ -->
<!-- 🕮 Read our guide about submitting issues: https://github.com/microsoft/vscode/wiki/Submitting-Bugs-and-Suggestions -->
<!-- 🔎 Search existing issues to avoid creating duplicates. -->
<!-- 🧪 Test using the latest Insiders build to see if your issue has already been fixed: https://code.visualstudio.com/insiders/ -->
<!-- 💡 Instead of creating your report here, use 'Report Issue' from the 'Help' menu in VS Code to pre-fill useful information. -->
<!-- 🔧 Launch with `code --disable-extensions` to check. -->
Does this issue occur when all extensions are disabled?: Yes

<!-- 🪓 If you answered No above, use 'Help: Start Extension Bisect' from Command Palette to try to identify the cause. -->
<!-- 📣 Issues caused by an extension need to be reported directly to the extension publisher. The 'Help > Report Issue' dialog can assist with this. -->
Version: 1.99.0 (user setup)
Commit: 4437686ffebaf200fa4a6e6e67f735f3edf24ada
Date: 2025-04-02T21:35:19.530Z
Electron: 34.3.2
ElectronBuildId: 11161073
Chromium: 132.0.6834.210
Node.js: 20.18.3
V8: 13.2.152.41-electron.0
OS: Windows_NT x64 10.0.26100



Since the march 2025 update variable substitution in tasks.json is (partly?) not working anymore.
We are referencing `${config:build.cwd}` in tasks.json. `build.cwd` is defined in the code-workspace file (which is opened in VS-Code).


Minimal example:

tasks.json:
```
{
	// See https://go.microsoft.com/fwlink/?LinkId=733558
	// for the documentation about the tasks.json format
	"version": "2.0.0",
	"tasks": [{
			"label": "Windows Release",
			"type": "shell",
			"options": {
				"cwd": "${config:build.cwd}",
			},
			"command": "./make.bat",
			"presentation": {
				"clear": true
			},
			"group": "build"
		}
	]
}
```
code-workspace file:
```
{
	"folders": [
		{
			"path": "../"
		},
	],
	"settings": {
		"pathToProject": "${workspaceFolder}/Tests/Unit_Tests/Testcases_FTL",
		"build": {
			"windows": {
				"platform": "Windows",
				"run_profile": "utest",
				// ...
			},
			"target": {
				"platform": "XMaster5",
				"run_profile": "utest",
				// ...
			},
			"pathToBuild": "${config:pathToProject}/build",
			"cwd": "${config:pathToProject}",
			"sourceLocation": "${config:pathToProject}",
		}
	}
}
```

Terminal output on starting Windows Release Task:
```
 *  Executing task in folder ${config:pathToProject}: ./make.bat -j16 all 


 *  The terminal process failed to launch: Starting directory (cwd) "D:\Repos\XM5\main_bws\${workspaceFolder}/Tests/Unit_Tests/Testcases_FTL" does not exist. 
```


With version 1.98.2 it's working fine.

I am experiencing what appears to be the same issue. I installed the March 2025 update today, and since then I'm having issues trying to launch my tasks. Everything is in a `code-workspace` file. Here are some details:

```jsonc
// The task looks like this...
{
    "label": "Build Project",
    "command": "dotnet",
    "type": "shell",
    "args": [
        "build",
        "${config:_custom_dir}"
    ],
    "problemMatcher": "$msCompile"
}
```
```jsonc
// Higher up in the file I have something like this...
"folders": [
    {
        "path": "..\\some\\relative\\path",
        "name": "Repo/Project"
    }
    // More workspace folders...
],
"settings": {
    "_custom_workspace": "${workspaceFolder:Repo/Project}",
    "_custom_dir": "${config:_custom_workspace}\\some\\subpath"
    // More custom settings...
}
```

When trying to run the task I get this error in the terminal:
```
MSBUILD : error MSB1009: Project file does not exist.
Switch: \some\subpath

 *  The terminal process "C:\WINDOWS\System32\WindowsPowerShell\v1.0\powershell.exe -Command dotnet build ${workspaceFolder:Repo/Project}\some\subpath" terminated with exit code: 1.
```

--- a/src/vs/workbench/services/configurationResolver/common/configurationResolverExpression.ts
+++ b/src/vs/workbench/services/configurationResolver/common/configurationResolverExpression.ts
@@ -68,6 +68,11 @@ export class ConfigurationResolverExpression<T> implements IConfigurationResolve
 	private locations = new Map<string, IReplacementLocation>();
 	private root: T;
 	private stringRoot: boolean;
+	/**
+	 * Callbacks when a new replacement is made, so that nested resolutions from
+	 * `expr.unresolved()` can be fulfilled in the same iteration.
+	 */
+	private newReplacementNotifiers = new Set<(r: Replacement) => void>();
 
 	private constructor(object: T) {
 		// If the input is a string, wrap it in an object so we can use the same logic
@@ -180,10 +185,9 @@ export class ConfigurationResolverExpression<T> implements IConfigurationResolve
 		for (const [key] of Object.entries(obj)) {
 			this.parseString(obj, key, key, true);
 		}
-
 	}
 
-	private parseString(object: any, propertyName: string | number, value: string, replaceKeyName?: boolean): void {
+	private parseString(object: any, propertyName: string | number, value: string, replaceKeyName?: boolean, replacementPath?: string[]): void {
 		let pos = 0;
 		while (pos < value.length) {
 			const match = value.indexOf('${', pos);
@@ -192,18 +196,53 @@ export class ConfigurationResolverExpression<T> implements IConfigurationResolve
 			}
 			const parsed = this.parseVariable(value, match);
 			if (parsed) {
-				const locations = this.locations.get(parsed.replacement.id) || { locations: [], replacement: parsed.replacement };
-				locations.locations.push({ object, propertyName, replaceKeyName });
-				this.locations.set(parsed.replacement.id, locations);
 				pos = parsed.end + 1;
+				if (replacementPath?.includes(parsed.replacement.id)) {
+					continue;
+				}
+
+				const locations = this.locations.get(parsed.replacement.id) || { locations: [], replacement: parsed.replacement };
+				const newLocation: PropertyLocation = { object, propertyName, replaceKeyName };
+				locations.locations.push(newLocation);
+				this.locations.set(parsed.replacement.id, locations);
+
+				if (locations.resolved) {
+					this._resolveAtLocation(parsed.replacement, newLocation, locations.resolved, replacementPath);
+				} else {
+					this.newReplacementNotifiers.forEach(n => n(parsed.replacement));
+				}
 			} else {
 				pos = match + 2;
 			}
 		}
 	}
 
-	public unresolved(): Iterable<Replacement> {
-		return Iterable.map(Iterable.filter(this.locations.values(), l => l.resolved === undefined), l => l.replacement);
+	public *unresolved(): Iterable<Replacement> {
+		const newReplacements = new Map<string, Replacement>();
+		const notifier = (replacement: Replacement) => {
+			newReplacements.set(replacement.id, replacement);
+		};
+
+		for (const location of this.locations.values()) {
+			if (location.resolved === undefined) {
+				newReplacements.set(location.replacement.id, location.replacement);
+			}
+		}
+
+		this.newReplacementNotifiers.add(notifier);
+
+		while (true) {
+			const next = Iterable.first(newReplacements);
+			if (!next) {
+				break;
+			}
+
+			const [key, value] = next;
+			yield value;
+			newReplacements.delete(key);
+		}
+
+		this.newReplacementNotifiers.delete(notifier);
 	}
 
 	public resolved(): Iterable<[Replacement, IResolvedValue]> {
@@ -220,22 +259,36 @@ export class ConfigurationResolverExpression<T> implements IConfigurationResolve
 			return;
 		}
 
+		location.resolved = data;
+
 		if (data.value !== undefined) {
-			for (const { object, propertyName, replaceKeyName } of location.locations || []) {
-				if (replaceKeyName && typeof propertyName === 'string') {
-					// replace key
-					const value = object[propertyName];
-					const newValue = propertyName.replaceAll(replacement.id, data.value);
-					delete object[propertyName];
-					object[newValue] = value;
-				} else {
-					const newValue = object[propertyName].replaceAll(replacement.id, data.value);
-					object[propertyName] = newValue;
-				}
+			for (const l of location.locations || Iterable.empty()) {
+				this._resolveAtLocation(replacement, l, data);
 			}
 		}
+	}
 
-		location.resolved = data;
+	private _resolveAtLocation(replacement: Replacement, { replaceKeyName, propertyName, object }: PropertyLocation, data: IResolvedValue, path: string[] = []) {
+		if (data.value === undefined) {
+			return;
+		}
+
+		// avoid recursive resolution, e.g. ${env:FOO} -> ${env:BAR}=${env:FOO}
+		path.push(replacement.id);
+
+		// note: in nested `this.parseString`, parse only the new substring for any replacements, don't reparse the whole string
+		if (replaceKeyName && typeof propertyName === 'string') {
+			const value = object[propertyName];
+			const newKey = propertyName.replaceAll(replacement.id, data.value);
+			delete object[propertyName];
+			object[newKey] = value;
+			this.parseString(object, newKey, data.value, true, path);
+		} else {
+			this.parseString(object, propertyName, data.value, false, path);
+			object[propertyName] = object[propertyName].replaceAll(replacement.id, data.value);
+		}
+
+		path.pop();
 	}
 
 	public toObject(): T {

--- a/src/vs/workbench/services/configurationResolver/test/electron-sandbox/configurationResolverService.test.ts
+++ b/src/vs/workbench/services/configurationResolver/test/electron-sandbox/configurationResolverService.test.ts
@@ -317,6 +317,17 @@ suite('Configuration Resolver Service', () => {
 		}
 	});
 
+	test('recursively resolve variables', async () => {
+		const configurationService = new TestConfigurationService({
+			key1: 'key1=${config:key2}',
+			key2: 'key2=${config:key3}',
+			key3: 'we did it!',
+		});
+
+		const service = new TestConfigurationResolverService(nullContext, Promise.resolve(envVariables), disposables.add(new TestEditorServiceWithActiveEditor()), configurationService, mockCommandService, new TestContextService(), quickInputService, labelService, pathService, extensionService, disposables.add(new TestStorageService()));
+		assert.strictEqual(await service.resolveAsync(workspace, '${config:key1}'), 'key1=key2=we did it!');
+	});
+
 	test('substitute many env variable and a configuration variable', async () => {
 		const configurationService = new TestConfigurationService({
 			editor: {
@@ -958,5 +969,25 @@ suite('ConfigurationResolverExpression', () => {
 		assert.strictEqual(unresolved.length, 1);
 		assert.strictEqual(unresolved[0].name, 'env');
 		assert.strictEqual(unresolved[0].arg, 'HOME${env:USER}');
+	});
+
+	test('resolves nested values', () => {
+		const expr = ConfigurationResolverExpression.parse({
+			name: '${env:REDIRECTED}',
+			'key that is ${env:REDIRECTED}': 'cool!',
+		});
+
+		for (const r of expr.unresolved()) {
+			if (r.arg === 'REDIRECTED') {
+				expr.resolve(r, 'username: ${env:USERNAME}');
+			} else if (r.arg === 'USERNAME') {
+				expr.resolve(r, 'testuser');
+			}
+		}
+
+		assert.deepStrictEqual(expr.toObject(), {
+			name: 'username: testuser',
+			'key that is username: testuser': 'cool!'
+		});
 	});
 });

